File: cpp/src/examples/gloo_example.cpp
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>

#include "cylon/net/gloo/gloo_communicator.hpp"
#include "cylon/ctx/cylon_context.hpp"

#include "examples/example_utils.hpp"

static constexpr int kCount = 10;
static constexpr double kDup = 0.9;

int main(int argc, char **argv) {
  auto config = std::make_shared<cylon::net::GlooConfig>(std::stoi(argv[1]), std::stoi(argv[2]));
  config->SetFileStorePath("/tmp/gloo/");
  config->SetStorePrefix("foo");

  if (config->rank() == 0)  {
    system("rm -rf /tmp/gloo/*");
  }

  std::shared_ptr<cylon::CylonContext> ctx;
  if (!cylon::CylonContext::InitDistributed(config, &ctx).is_ok()) {
    return 1;
  }

  LOG(INFO) << "rank:" << ctx->GetRank() << " size:" << ctx->GetWorldSize();

  std::shared_ptr<cylon::Table> first_table, second_table, out;

  cylon::examples::create_two_in_memory_tables(kCount, kDup, ctx, first_table, second_table);

  cylon::join::config::JoinConfig jc{cylon::join::config::JoinType::INNER, 0, 0,
                                     cylon::join::config::JoinAlgorithm::SORT, "l_", "r_"};

  auto status = cylon::DistributedJoin(first_table, second_table, jc, out);

  if (!status.is_ok()) {
    LOG(INFO) << "Table join failed ";
    return 1;
  }

  LOG(INFO) << "First table had : " << first_table->Rows() << " and Second table had : "
            << second_table->Rows() << ", Joined has : " << out->Rows();
  return 0;
}

File: python/pycylon/examples/gloo_example.py
#!/usr/bin/env python

#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# import subprocess
#
import os
import random
import tempfile
from multiprocessing import Process

from pycylon import CylonEnv, DataFrame
from pycylon.net.gloo_config import GlooStandaloneConfig

FILE_STORE_PATH = os.path.join(tempfile.gettempdir(), 'gloo')
WORLD_SIZE = 4


def run_op(env):
    df1 = DataFrame([random.sample(range(10 * env.rank, 15 * (env.rank + 1)), 5),
                     random.sample(range(10 * env.rank, 15 * (env.rank + 1)), 5)])
    df2 = DataFrame([random.sample(range(10 * env.rank, 15 * (env.rank + 1)), 5),
                     random.sample(range(10 * env.rank, 15 * (env.rank + 1)), 5)])
    print("Distributed Merge")
    df3 = df1.merge(right=df2, on=[0], env=env)
    print(df3)

    env.finalize()


def run_standalone_gloo(rank, world_size):
    conf = GlooStandaloneConfig(rank, world_size)

    if not os.path.isdir(FILE_STORE_PATH):
        os.mkdir(FILE_STORE_PATH)
    conf.set_file_store_path(FILE_STORE_PATH)

    # distributed join
    env = CylonEnv(config=conf)
    run_op(env)


if __name__ == "__main__":  # confirms that the code is under main function
    procs = []

    # instantiating process with arguments
    for i in range(WORLD_SIZE):
        # print(name)
        proc = Process(target=run_standalone_gloo, args=(i, WORLD_SIZE))
        procs.append(proc)
        proc.start()

    # complete the processes
    for proc in procs:
        proc.join()
